dev/clint/src/clint/linter.py

```diff
@@ -0,0 +1,170 @@
+        return f"{self.path}:{self.lineno}:{self.col_offset}: {self.rule.id}: {self.rule.message}"
+
+    def json(self) -> dict[str, str | int | None]:
+        return {
+            "type": "error",
+            "module": None,
+            "obj": None,
+            "line": self.lineno,
+            "column": self.col_offset,
+            "endLine": self.lineno,
+            "endColumn": self.col_offset,
+            "path": self.path,
+            "symbol": self.rule.name,
+            "message": self.rule.message,
+            "message-id": self.rule.id,
+        }
+
+
+NO_RST = Rule(
+    "Z0001",
+    "no-rst",
+    "Do not use RST style. Use Google style instead.",
+)
+LAZY_BUILTIN_IMPORT = Rule(
+    "Z0002",
+    "lazy-builtin-import",
+    "Builtin modules must be imported at the top level.",
+)
+
+# TODO: Remove this once we convert all docstrings to Google style.
+NO_RST_IGNORE = {
+    "mlflow/gateway/client.py",
+    "mlflow/gateway/providers/utils.py",
+    "mlflow/keras/callback.py",
+    "mlflow/metrics/base.py",
+    "mlflow/metrics/genai/base.py",
+    "mlflow/models/utils.py",
+    "mlflow/projects/databricks.py",
+    "mlflow/projects/kubernetes.py",
+    "mlflow/store/_unity_catalog/registry/rest_store.py",
+    "mlflow/store/artifact/azure_data_lake_artifact_repo.py",
+    "mlflow/store/artifact/gcs_artifact_repo.py",
+    "mlflow/store/model_registry/rest_store.py",
+    "mlflow/store/tracking/rest_store.py",
+    "mlflow/utils/docstring_utils.py",
+    "mlflow/utils/rest_utils.py",
+    "tests/utils/test_docstring_utils.py",
+}
+
+
+class Linter(ast.NodeVisitor):
+    def __init__(self, path: str, ignore: dict[str, set[int]]):
+        self.stack: list[ast.FunctionDef | ast.AsyncFunctionDef] = []
+        self.path = path
+        self.ignore = ignore
+        self.violations: list[Violation] = []
+
+    def _check(self, node: ast.AST, rule: Rule) -> None:
+        if (lines := self.ignore.get(rule.name)) and node.lineno in lines:
+            return
+        self.violations.append(Violation(rule, self.path, node.lineno, node.col_offset))
+
+    def _docstring(
+        self, node: ast.FunctionDef | ast.AsyncFunctionDef | ast.ClassDef
+    ) -> ast.Constant | None:
+        if (
+            isinstance(node.body[0], ast.Expr)
+            and isinstance(node.body[0].value, ast.Constant)
+            and isinstance(node.body[0].value.s, str)
+        ):
+            return node.body[0].value
+        return None
+
+    def _no_rst(self, node: ast.FunctionDef | ast.AsyncFunctionDef) -> None:
+        if self.path in NO_RST_IGNORE:
+            return
+
+        if (nd := self._docstring(node)) and (
+            PARAM_REGEX.search(nd.s) or RETURN_REGEX.search(nd.s)
+        ):
+            self._check(nd, NO_RST)
+
+    def _is_in_function(self) -> bool:
+        return bool(self.stack)
+
+    def visit_ClassDef(self, node: ast.ClassDef) -> None:
+        self.stack.append(node)
+        self._no_rst(node)
```

Comment:
This is more about future extension, but is it possible to define those rules pluggable like pylint (or perhaps declarative)? Currently when we want to add/remove rules for clint, we need to directly edit the single node traversal class, which might be a bit error-prone. Of course it doesn't need to be in this PR but sth we can think of in long-term.

Reply:
Defining a visitor for each rule like pylint has a con. It's slow because _**each**_ visitor needs to traverse the code. I'd consider making this pluggable if we had more (> 5) rules.